Normalize late-bound projections with inference variables in fulfillment

compiler-errors opened the pull request:
I modified the PR #90887 <i>"Try to normalize associated types before processing obligations"</i> to only attempt normalization when these two conditions are both true:
1. the projection type has escaping late-bound variables, and 
2. the projection has remaining inference variables.

The only projections that we expect to see inside of a predicate during fulfillment (and which we have any hope of normalizing further) are those that have remaining inference variables that need to be resolved, and which have late-bound vars which opted them out of being replaced with a fresh inference variable in the first place. We rely on existing normalization that is speckled through typeck and confirmation (e.g. [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_trait_selection/src/traits/select/mod.rs#L2339)) to uphold this invariant, which means we can skip a large portion of the normalization calls that we saw in #90887.

As seen in the (really messy) initial PR I put up for a perf run, it [seems to have improved performance](https://github.com/rust-lang/rust/pull/94042#issuecomment-1041303420).

----

Reasoning about why these two conditions together are sufficient: 
1. If we see a projection during the fulfillment loop that has inference but no late-bound variables, we probably are missing a normalization call somewhere else, since that projection should've been replaced with an inference variable when we got the type from HIR. This means we should track down that missing normalization call to somewhere in typeck or confirmation.
2. If we see a projection with escaping late-bound variables (but no inference variables), then we probably don't have a solution to that projection variable in the first place, since we should have been able to normalize it (since we have all the information we can have about the type).

This probably needs more discussion, but I'm putting this up for initial review, and we can perhaps discuss it in the GATs meeting.

r? @nikomatsakis 
cc @jackh726 

Fixes #90729
Fixes #93341
Fixes #93342
Fixes #94160

jackh726 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit df8d9f530e6a6c825f92b21145861dbbabeb083e with merge f20c37cb312b7df326fa1d05e51af9149c90e9fd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"df8d9f530e6a6c825f92b21145861dbbabeb083e","merge_sha":"f20c37cb312b7df326fa1d05e51af9149c90e9fd"} -->

jackh726 commented:
Things to think about:
* I liked removing the extra normalizations in confirmations; it really highlights that this actually simplifies things (this also in theory could have been the reason for better perf, we'll see)
* Was moving the normalization in fulfill up a bit enough to fix perf? In other words, do we *really* need to only normalize if there are late-bound vars?

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5229141922?check_suite_focus=true)
Build commit: f20c37cb312b7df326fa1d05e51af9149c90e9fd (`f20c37cb312b7df326fa1d05e51af9149c90e9fd`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5229141922?check_suite_focus=true"},"merge_sha":"f20c37cb312b7df326fa1d05e51af9149c90e9fd"} -->

rust-timer commented:
Queued f20c37cb312b7df326fa1d05e51af9149c90e9fd with parent 930fc4f59ddeb9f26d554a2c75c5355989189540, future [comparison URL](https://perf.rust-lang.org/compare.html?start=930fc4f59ddeb9f26d554a2c75c5355989189540&end=f20c37cb312b7df326fa1d05e51af9149c90e9fd).

compiler-errors commented:
Yeah, regard the first part, I actually discovered that all that extra normalizing of those Fn trait refs wasn't needed when I stopped first shallow resolving and then normalizing during fulfillment... hence this PR is a lot shorter.

The second question is a good point, I can put up a quick PR tomorrow unless you want to modify #90887 yourself.

rust-timer commented:
Finished benchmarking commit (f20c37cb312b7df326fa1d05e51af9149c90e9fd): [comparison url](https://perf.rust-lang.org/compare.html?start=930fc4f59ddeb9f26d554a2c75c5355989189540&end=f20c37cb312b7df326fa1d05e51af9149c90e9fd).

**Summary**: This benchmark run shows 13 relevant regressions 😿 to instruction counts.
- Average relevant regression: 1.3%
- Largest regression in instruction counts: 3.1% on `full` builds of `inflate check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

compiler-errors commented:
Haha! That's lovely. Let me rebase that normalization commit when I get up in a few hours.

jackh726 commented:
@bors try @rust-timer queue

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5236188232?check_suite_focus=true)
Build commit: 0ab1cb526af76f0c1d83259e21fd9cb9497c1e80 (`0ab1cb526af76f0c1d83259e21fd9cb9497c1e80`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5236188232?check_suite_focus=true"},"merge_sha":"0ab1cb526af76f0c1d83259e21fd9cb9497c1e80"} -->

rust-timer commented:
Queued 0ab1cb526af76f0c1d83259e21fd9cb9497c1e80 with parent 30b3f35c420694a4f24e5a4df00f06073f4f3a37, future [comparison URL](https://perf.rust-lang.org/compare.html?start=30b3f35c420694a4f24e5a4df00f06073f4f3a37&end=0ab1cb526af76f0c1d83259e21fd9cb9497c1e80).

rust-timer commented:
Finished benchmarking commit (0ab1cb526af76f0c1d83259e21fd9cb9497c1e80): [comparison url](https://perf.rust-lang.org/compare.html?start=30b3f35c420694a4f24e5a4df00f06073f4f3a37&end=0ab1cb526af76f0c1d83259e21fd9cb9497c1e80).

**Summary**: This benchmark run did not return any relevant results. 14 results were found to be statistically significant but too small to be relevant.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

compiler-errors commented:
Interesting to see that the perf improvement is a combination of the normalization in selection _and_ (probably?) the type flag addition.

Per discussion with @jackh726, let's test this hypothesis further. I reverted the type flag diff (and added a small change so we don't resolve variables _and_ normalize in fulfillment, since that's kinda doing duplicated work).

jackh726 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit a2b03db0ad3ef211d84a8348b5bec6854d7d8ee5 with merge f9731aa0252167d632138fe7bcbfe90f406178f6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a2b03db0ad3ef211d84a8348b5bec6854d7d8ee5","merge_sha":"f9731aa0252167d632138fe7bcbfe90f406178f6"} -->

jackh726 commented on compiler/rustc_trait_selection/src/traits/fulfill.rs:
```diff
@@ -344,17 +344,32 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
+                obligations.push(obligation.with(predicate));
+                return ProcessResult::Changed(mk_pending(obligations));
+            }
+        } else if obligation.predicate.has_infer_types_or_consts() {
```
Just fyi this now skips the inference var resolution if there is a projection but we can't normalize.

compiler-errors replied:
I was thinking about that just now... Do you want me to revert and we can re-queue the test, or should we just do perf on whatever's currently queued?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5238078130?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a2b03db0ad3ef211d84a8348b5bec6854d7d8ee5/checks/5238078130/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
normalized stderr:
error[E0311]: the parameter type `T` may not live long enough
  --> $DIR/issue-91139.rs:17:12
   |
LL | fn foo<T>() {
   |        - help: consider adding an explicit lifetime bound...: `T: 'a`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
LL |     let _: for<'a> fn(<() as Foo<T>>::Type<'a>, &'a T) = |_, _| ();
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `T` will meet its required lifetime bounds
error: aborting due to previous error



---
To only update this specific test, also pass `--test-args generic-associated-types/issue-91139.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/generic-associated-types/issue-91139.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/issue-91139" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/issue-91139/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0311]: the parameter type `T` may not live long enough
  --> /checkout/src/test/ui/generic-associated-types/issue-91139.rs:17:12
   |
LL | fn foo<T>() {
   |        - help: consider adding an explicit lifetime bound...: `T: 'a`
LL |     let _: for<'a> fn(<() as Foo<T>>::Type<'a>, &'a T) = |_, _| ();
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `T` will meet its required lifetime bounds
error: aborting due to previous error


------------------------------------------
```

</details>

jackh726 commented:
@bors try-

jackh726 commented:
@bors try

bors commented:
:hourglass: Trying commit e92dac75d24d3bfc18174692055e966b87e15ce7 with merge b133e1a0ceb78dbf55643a6deab904d3911208b6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e92dac75d24d3bfc18174692055e966b87e15ce7","merge_sha":"b133e1a0ceb78dbf55643a6deab904d3911208b6"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5239781855?check_suite_focus=true)
Build commit: b133e1a0ceb78dbf55643a6deab904d3911208b6 (`b133e1a0ceb78dbf55643a6deab904d3911208b6`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5239781855?check_suite_focus=true"},"merge_sha":"b133e1a0ceb78dbf55643a6deab904d3911208b6"} -->

rust-timer commented:
Queued b133e1a0ceb78dbf55643a6deab904d3911208b6 with parent 30b3f35c420694a4f24e5a4df00f06073f4f3a37, future [comparison URL](https://perf.rust-lang.org/compare.html?start=30b3f35c420694a4f24e5a4df00f06073f4f3a37&end=b133e1a0ceb78dbf55643a6deab904d3911208b6).

bors commented:
:umbrella: The latest upstream changes (presumably #90887) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 0a096a0c82978f64e8bb0e3b2923e2f954b4cb6f with merge dd0319c25a100e481bc99429d2a54fe68330aa26...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0a096a0c82978f64e8bb0e3b2923e2f954b4cb6f","merge_sha":"dd0319c25a100e481bc99429d2a54fe68330aa26"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5461096057?check_suite_focus=true)
Build commit: dd0319c25a100e481bc99429d2a54fe68330aa26 (`dd0319c25a100e481bc99429d2a54fe68330aa26`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5461096057?check_suite_focus=true"},"merge_sha":"dd0319c25a100e481bc99429d2a54fe68330aa26"} -->

rust-timer commented:
Queued dd0319c25a100e481bc99429d2a54fe68330aa26 with parent 67b3e8183830c7af4e06a9aa91de4d1be3c860f7, future [comparison URL](https://perf.rust-lang.org/compare.html?start=67b3e8183830c7af4e06a9aa91de4d1be3c860f7&end=dd0319c25a100e481bc99429d2a54fe68330aa26).

rust-timer commented:
Finished benchmarking commit (dd0319c25a100e481bc99429d2a54fe68330aa26): [comparison url](https://perf.rust-lang.org/compare.html?start=67b3e8183830c7af4e06a9aa91de4d1be3c860f7&end=dd0319c25a100e481bc99429d2a54fe68330aa26).

**Summary**: This benchmark run shows 15 relevant improvements 🎉 to instruction counts.
- Arithmetic mean of relevant improvements: -0.4%
- Largest improvement in instruction counts: -0.8% on `full` builds of `futures check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

nikomatsakis commented:
@compiler-errors do you plan to rebase atop https://github.com/rust-lang/rust/pull/90887 ?

jackh726 commented:
Not sure the slight perf improvement here is worth the added complexity.

compiler-errors commented:
Yeah, in that case, I'll close this out.
